src/components/non-bitgo.js

```diff
@@ -105,84 +116,108 @@ class NonBitGoRecoveryForm extends Component {
     }
   };
 
-  async performRecovery() {
-    this.setState({ recovering: true, error: '' });
+  getRecoveryParams() {
+    // This is like _.pick
+    return [
+      'userKey',
+      'backupKey',
+      'bitgoKey',
+      'rootAddress',
+      'walletContractAddress',
+      'tokenAddress',
+      'walletPassphrase',
+      'recoveryDestination',
+      'scan',
+      'krsProvider',
+    ].reduce((obj, param) => {
+      if (this.state[param]) {
+        const value = this.state[param];
 
-    const baseCoin = await this.getCoinObject();
+        return Object.assign(obj, { [param]: value });
+      }
+      return obj;
+    }, {});
+  }
 
-    const recoveryTool = baseCoin.recover;
+  async performRecoveryWithParams(baseCoin, recoveryParams) {
+    if ((this.state.coin === 'bsv' || this.state.coin === 'bch' || this.state.coin === 'bcha') && this.state.apiKey) {
+      recoveryParams.apiKey = this.state.apiKey;
+    }
 
-    if (!recoveryTool) {
-      this.setState({ error: `Recovery tool not found for ${this.state.coin}`, recovering: false });
+    if (!coinConfig.allCoins[this.state.coin].recoverP2wsh) {
+      recoveryParams.ignoreAddressTypes = ['p2wsh'];
+    }
+
+    const recovery = await recoverWithKeyPath(baseCoin, recoveryParams);
+
+    const recoveryTx = recovery.transactionHex || recovery.txHex || recovery.tx;
+
+    if (!recoveryTx) {
+      throw new Error('Fully-signed recovery transaction not detected.');
+    }
+
+    const fileName = baseCoin.getChain() + '-recovery-' + Date.now().toString() + '.json';
+    const dialogParams = {
+      filters: [
+        {
+          name: 'Custom File Type',
+          extensions: ['json'],
+        },
+      ],
+      defaultPath: '~/' + fileName,
```

Comment:
Might wanna use `os.homedir()` for this: https://nodejs.dev/learn/the-nodejs-os-module#oshomedir
Not sure if `~/` is easy for users to find on every operating system (looking at you, Windows!)

Reply:
You are correct, but I am just moving around existing code here. We should tackle these sorts of issues in separate PRs.

Reply:
👍